ocaml/xapi/xapi_host.ml

```diff
@@ -2246,20 +2246,36 @@ let migrate_receive ~__context ~host ~network ~options =
            ( Api_errors.host_cannot_attach_network
            , [Ref.string_of host; Ref.string_of network] ))
   in
-  let ip = Db.PIF.get_IP ~__context ~self:pif in
-  ( if String.length ip = 0 then
-      match Db.PIF.get_ip_configuration_mode ~__context ~self:pif with
-      | `None ->
-          raise
-            (Api_errors.Server_error
-               (Api_errors.pif_has_no_network_configuration, [Ref.string_of pif]))
-      | `DHCP ->
-          raise
-            (Api_errors.Server_error
-               (Api_errors.interface_has_no_ip, [Ref.string_of pif]))
-      | _ ->
-          failwith "No IP address on PIF"
+  let primary_address_type =
+    Db.PIF.get_primary_address_type ~__context ~self:pif
+  in
+  let ip, configuration_mode =
+    match primary_address_type with
+    | `IPv4 ->
+      ( Db.PIF.get_IP ~__context ~self:pif,
+      Db.PIF.get_ip_configuration_mode ~__context ~self:pif )
+    | `IPv6 ->
+      ( List.hd (Db.PIF.get_IPv6 ~__context ~self:pif),
```

Comment:
Avoid `List.hd`, which gives nasty exceptions. You could make `ip` a list, and just remove the `List.hd` here and add `[ ... ]` above.

Reply:
How can i get the first elem of a list "cleanly" then?

Reply:
Because later I need the value `ip` as a string to add the extra `[]` around IPv6 for URL.

Reply:
```
match primary_address_type, ip with
    | `IPv4, ip :: _ -> ip
    | `IPv6, ip :: _ -> Http.Url.maybe_wrap_IPv6_literal ip
```

Reply:
what if I do this:

```ocaml
  let ip, configuration_mode =
    match primary_address_type with
    | `IPv4 ->
      ( Db.PIF.get_IP ~__context ~self:pif,
      Db.PIF.get_ip_configuration_mode ~__context ~self:pif )
    | `IPv6 ->
      ( match Db.PIF.get_IPv6 ~__context ~self:pif with
         | [] -> ""
         | hd::tl -> hd,
      Db.PIF.get_ipv6_configuration_mode ~__context ~self:pif )
  in
```

Reply:
That's fine as well.

Reply:
Yours is more concise, i'll do that.